### DISCUSSION TITLE
Update mesh on properties

### DISCUSSION THREAD

**Discussion by PBrockmann** [pyvista/pyvista#2567] [Q&A]
How to update the display of a mesh on some properties, for example show_edges ?
Remove the actor and recreate it ?
Is it efficient ?

Here is what I have implemented as a demo ?

```python
import pyvista as pv
from ipywidgets import interact

pl = pv.Plotter()
actor = pl.add_mesh(pv.Sphere(), show_edges=True)

def plot(show_edges=True):
    global actor
    pl.remove_actor(actor)
    actor = pl.add_mesh(pv.Sphere(), show_edges=show_edges)
    pl.show(jupyter_backend='pythreejs')
    
a = interact(plot, show_edges=True)
```

**Comment by PBrockmann** [comment]
Also how to keep position of the camera ?
Relative to question: https://github.com/pyvista/pyvista/discussions/2562

**Comment by ndahiya3** [comment]
Removing the actor and re-adding it with the updated property (with updated show_edges, color or opacity) kinda works if you have to do it occasionally. I think you can kind of see when a mesh is removed and re-added, it sort of seems to jump. I think if you have do it continuously (say continuously change opacity with a slider) then it'll probably look odd. I think some properties can be changed by using vtk. I haven't tried changing things continuously with a slider yet but you can try the following:

`actor.GetProperty().SetColor([1,0,0]) # Set color red`
`actor.GetProperty().SetEdgeVisibility(True) # or False`
`actor.GetProperty().SetOpacity(0.5) # Set opacity, could be continuously updated with a slider`
`actor.SetVisibility(False) # or True`

**Comment by PBrockmann** [comment]
Ok for changing a property of the actor.
But how to update the rendering ?
I still have the "jump" when I toggle the "show_edges" with the following lines.

```
import pyvista as pv
from ipywidgets import interact

pl = pv.Plotter()
actor = pl.add_mesh(pv.Sphere(), show_edges=True)

def plot(show_edges=True):
    global actor
    actor.GetProperty().SetEdgeVisibility(show_edges)
    pl.show(jupyter_backend='pythreejs')
    #pl.update()    # does no work complains about "Render window interactor must be initialized before processing events."
    
a = interact(plot, show_edges=True)
```

**Comment by ndahiya3** [comment]
I can't seem to run your example. I tried the following. To be honest I don't see much difference between setting vtk property vs removing and re-adding the mesh. But removing/adding approach will make the mesh disappear for a fraction of a second and you might notice that blinking. In the other case, it does take a fraction of a second to show the edges and the mesh appearance changes a bit suddenly.
```
import sys
# Setting the Qt bindings for QtPy
import os
os.environ["QT_API"] = "pyside2"
#os.environ["QT_API"] = "pyqt5"

from qtpy import QtWidgets

import numpy as np

import pyvista as pv
from pyvistaqt import QtInteractor, MainWindow
from pyvista import examples
bunny = examples.download_bunny()

class MyMainWindow(MainWindow):

    def __init__(self, parent=None, show=True):
        QtWidgets.QMainWindow.__init__(self, parent)

        # create the frame
        self.frame = QtWidgets.QFrame()
        vlayout = QtWidgets.QVBoxLayout()

        # add the pyvista interactor object
        self.plotter = QtInteractor(self.frame)
        vlayout.addWidget(self.plotter.interactor)
        self.signal_close.connect(self.plotter.close)

        self.frame.setLayout(vlayout)
        self.setCentralWidget(self.frame)

        # simple menu to demo functions
        mainMenu = self.menuBar()
        fileMenu = mainMenu.addMenu('File')
        exitButton = QtWidgets.QAction('Exit', self)
        exitButton.setShortcut('Ctrl+Q')
        exitButton.triggered.connect(self.close)
        fileMenu.addAction(exitButton)

        # allow adding a sphere
        meshMenu = mainMenu.addMenu('Mesh')
        
        self.toggle_edges_vtk_action = QtWidgets.QAction('Toggle Edges VTK', self)
        self.toggle_edges_vtk_action.triggered.connect(self.toggle_edges_vtk)
        meshMenu.addAction(self.toggle_edges_vtk_action)
        
        self.toggle_edges_pyvista_action = QtWidgets.QAction('Toggle Edges Vista', self)
        self.toggle_edges_pyvista_action.triggered.connect(self.toggle_edges_pyvista)
        meshMenu.addAction(self.toggle_edges_pyvista_action)
        
        self.show_edges = False
#        self.sphere = pv.Sphere()
        self.sphere = bunny
        self.actor = self.plotter.add_mesh(self.sphere, show_edges=self.show_edges)
        
        if show:
            self.show()

    def toggle_edges_vtk(self):
        """ Toggle Edges of sphere """
        self.show_edges = not self.show_edges
        self.actor.GetProperty().SetEdgeVisibility(self.show_edges)
        
    def toggle_edges_pyvista(self):
        """ Toggle Edges of sphere by removing and re adding the actor"""
        self.show_edges = not self.show_edges
        self.plotter.remove_actor(self.actor)
        self.actor = self.plotter.add_mesh(self.sphere, show_edges=self.show_edges)


if __name__ == '__main__':
    app = QtWidgets.QApplication(sys.argv)
    window = MyMainWindow()
    sys.exit(app.exec_())
```

**Comment by PBrockmann** [comment]
Here is a solution to run in a notebook. Can it be simplified ? 

```
import vtk
vtk.vtkLogger.SetStderrVerbosity(vtk.vtkLogger.VERBOSITY_OFF)

import pyvista as pv
from ipywidgets import interact

pl = pv.Plotter()
actor = pl.add_mesh(pv.Sphere(), show_edges=True)
viewer = pl.show(jupyter_backend='pythreejs', return_viewer=True)

def plot(show_edges):
    global actor, viewer
    pl.remove_actor(actor)
    actor = pl.add_mesh(pv.Sphere(), show_edges=show_edges)
    viewer = pl.show(jupyter_backend='pythreejs', return_viewer=True)
    display(viewer)
    
a = interact(plot, show_edges=True)
```

More generally I am looking for help on how to update a viewer with ipywidget interactions.
I have tested pythreejs and panel backends and the code have to be modified  (saving the position of the camera from the viewer with pythreejs is not needed with the panel backend for exemple).
Do I have to remove actors each time a widget is changed ? Doing this with a panel backend leads to a window that blinks since the scene is redrawn each time.
Sure that I might have misunderstood some things...

So it would be nice to have some feedbacks on those tries.
https://github.com/PBrockmann/LSCE_notebooks/blob/main/pyvista_DYNAMICO_orthographic.ipynb

![image](https://user-images.githubusercontent.com/5402758/204063269-4777b6a0-a2c7-4b33-af08-4c7f9489205b.png)

**Comment by PBrockmann** [comment]
Have just tested the trame backend.

How to update a parameter without having to remove the actor each time ?


```
import pyvista as pv
from ipywidgets import interact

pv.set_jupyter_backend('trame')

pl = pv.Plotter()
c = pv.Cone(resolution=6)
c_actor = pl.add_mesh(c)
pl.show(jupyter_backend='client')

def plot(resolution=6):
    global c_actor
    pl.remove_actor(c_actor)
    c = pv.Cone(resolution=resolution)
    c_actor = pl.add_mesh(c)

a = interact(plot, resolution = (3,20,1))
```

**Reply by banesullivan** [reply]
Like this with a shallow copy of the mesh already added to the scene.

```py
import pyvista as pv
from ipywidgets import interact

pv.set_jupyter_backend('trame')

pl = pv.Plotter()
c = pv.Cone(resolution=6)
c_actor = pl.add_mesh(c)
pl.show(jupyter_backend='client')

def plot(resolution=6):
    c.shallow_copy(pv.Cone(resolution=resolution))
    pl.render()

a = interact(plot, resolution = (3,20,1))
```

then you can later change the properties:

```py
c_actor.prop.color = 'red'
pl.render()
```

The `render()` calls are necessary.

**Reply by banesullivan** [reply]
FWIW, you may want to check out using vtkAlgorithms: https://docs.pyvista.org/examples/99-advanced/plotting-algorithms.html#sphx-glr-examples-99-advanced-plotting-algorithms-py

**Comment by PBrockmann** [comment]
Thank you for pointing the use of shallow_copy function.
I have found clearer the update of the vtk object itself.

```
import vtk
import pyvista as pv
from ipywidgets import interact

pv.set_jupyter_backend('trame')

pl = pv.Plotter()
#c = pv.Cone(resolution=6)
c = vtk.vtkConeSource()
c_actor = pl.add_mesh(c)
pl.show(jupyter_backend='client')

def plot(resolution=6):
    #c.shallow_copy(pv.Cone(resolution=resolution))
    c.SetResolution(resolution)
    pl.render()

a = interact(plot, resolution = (3,20,1))
```

What about the update of a mesh, do I still need to remove it and add it again ?

**Reply by banesullivan** [reply]
> What about the update of a mesh, do I still need to remove it and add it again ?

You shouldn't need to

**Comment by PBrockmann** [comment]
I have updated my test notebook.

Am I missing efficiency by removing each time the mesh ?
I haven't been able to do otherwise.

![image](https://user-images.githubusercontent.com/5402758/218813358-9964ad67-0891-4ad4-94cd-d2cf48be2deb.png)

https://github.com/PBrockmann/LSCE_notebooks/blob/main/pyvista_DYNAMICO_orthographic_trame.ipynb
Updated

**Reply by banesullivan** [reply]
> Am I missing efficiency by removing each time the mesh ?

yes

**Reply by PBrockmann** [reply]
Let me know....

**Reply by PBrockmann** [reply]
I have tried to use ```pl.update_scalars(var, mesh=mesh)``` without success.

What are the hints ?

**Reply by banesullivan** [reply]
Just update the scalars on the mesh object itself (`mesh['name'] = var`) and call `plotter.render()` should be all that's needed

**Reply by PBrockmann** [reply]
Have been able to replace:

```mesh_actor = pl.add_mesh(mesh, scalars=var, show_edges=show_edges, clim=clim, cmap=cmap, show_scalar_bar=False)```
by
```
mesh_actor.prop.show_edges = show_edges
mesh_actor.mapper.set_scalars(var, '', cmap=cmap, clim=clim)
```

but it is not possible to get prop from the scalarBar_actor for updating as for mesh_actor

```scalarBar_actor = pl.add_scalar_bar(title=variable, color='black')```

```scalarBar_actor.prop``` does not exist ?

**Reply by PBrockmann** [reply]
Have encountered the bug https://github.com/pyvista/pyvista/issues/3615 when I have tried to use:
```pl.update_scalar_bar_range```
